components/tikv_util/src/config.rs

```diff
@@ -1456,68 +1433,115 @@ impl RaftDataStateMachine {
         Ok(())
     }
 
-    /// Enters the `Migrating` state and returns the source directory if a
-    /// migration is needed. Otherwise prepares the target directory for
-    /// opening.
-    pub fn before_open_target(&mut self) -> Option<PathBuf> {
+    /// Returns whether a migration is needed. When it's needed, enters the `Migrating`
+    /// state. Otherwise prepares the target directory for opening.
+    pub fn before_open_target(&mut self) -> bool {
         // Clean up trash directory if there is any.
-        for p in [
-            &self.source,
-            &self.source_staging,
-            &self.target,
-            &self.target_staging,
-        ] {
+        for p in [&self.source, &self.target] {
             let trash = p.with_extension("REMOVE");
             if trash.exists() {
                 fs::remove_dir_all(&trash).unwrap();
             }
         }
-        if Self::data_exists(&self.target_staging) {
-            // InvMigrating -> Completed
-            assert!(!Self::data_exists(&self.source_staging));
-            Self::remove_dir_safe(&self.source);
-            Self::rename_dir_safe(&self.target_staging, &self.target);
-            return None;
-        }
-        if Self::data_exists(&self.source) {
-            // Init -> Migrating
-            assert!(!Self::data_exists(&self.source_staging));
-            Self::rename_dir_safe(&self.source, &self.source_staging);
-        } else if !Self::data_exists(&self.source_staging) {
-            // No source data.
-            return None;
+        if !Self::data_exists(&self.source) {
+            // Recover from Completed state.
+            if self.in_progress_marker.exists() {
+                Self::must_remove(&self.in_progress_marker);
+            }
+            return false;
+        } else if self.in_progress_marker.exists() {
+            if let Some(real_source) = self.read_marker() {
+                // Recover from Migrating state.
+                if real_source == self.target {
+                    assert!(Self::data_exists(&self.target));
```

Comment:
I don't get it. Isn't it a valid case that migrate from target to source but panic before deleting the migrating success mark? I think it's just the same case of L1448 but in the reverse way.

Reply:
In that case, target will not be deleted until marker is deleted first.

Reply:
You are right, I'll check why unit test didn't cover this case.